xkafka/producer.go

```diff
@@ -0,0 +1,195 @@
+package xkafka
+
+import (
+	"context"
+	"strings"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/pkg/errors"
+
+	"github.com/gojekfarm/xtools/xkafka/internal"
+)
+
+// ProducerFunc is a function that returns a KafkaProducer.
+type ProducerFunc func(cfg *kafka.ConfigMap) (internal.ProducerClient, error)
+
+func (pf ProducerFunc) apply(o *options) { o.producerFn = pf }
+
+// DefaultProducerFunc is the default producer function that initialises
+// a new confluent-kafka-go/kafka.Producer.
+func DefaultProducerFunc(cfg *kafka.ConfigMap) (internal.ProducerClient, error) {
+	return kafka.NewProducer(cfg)
+}
+
+// Producer manages the production of messages to kafka topics.
+// It provides both synchronous and asynchronous publish methods
+// and a channel to stream delivery events.
+type Producer struct {
+	config              options
+	kafka               internal.ProducerClient
+	events              chan kafka.Event
+	delivery            chan *Message
+	middlewares         []middleware
+	wrappedPublish      Handler
+	wrappedAsyncPublish Handler
+}
+
+// NewProducer creates a new Producer.
+func NewProducer(name string, opts ...Option) (*Producer, error) {
+	cfg := defaultProducerOptions()
+
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	_ = cfg.configMap.SetKey("bootstrap.servers", strings.Join(cfg.brokers, ","))
+	_ = cfg.configMap.SetKey("client.id", name)
+
+	producer, err := cfg.producerFn(&cfg.configMap)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Producer{
+		config:   cfg,
+		kafka:    producer,
+		delivery: make(chan *Message),
+		events:   producer.Events(),
+	}
+
+	p.wrappedPublish = p.publish()
+	p.wrappedAsyncPublish = p.asyncPublish()
+
+	go p.start()
```

Comment:
`Start` and `Stop` should be exposed publicly IMO. Gives a user more fine grain control.

Reply:
https://github.com/gojekfarm/xtools/pull/7#discussion_r1130976137 making this change might remove the need for a public start & stop

Reply:
I see the value behind your point. However, when I was using xkafka in one of the services, I wanted to have finer control over when it starts and when it stops to test the start/stop process in unit-tests as well.

Perhaps, we can add an `Option` say
```go
type AutoStart bool
```
and default value can be true, however, it can still allow end-user to disable auto-start if needed.
WDYT?

Reply:
Also, I'm planning to use this open-source xkafka wrapper in another service which uses `xrun` heavily. If these components can also implement this [Component](https://pkg.go.dev/github.com/gojekfarm/xrun#Component) interface, that will be great.

Reply:
You may keep the main runner logic inside `Start(context.Context) error` as before, and add `Run` separately which calls `Start` first, then calls `Close` on exit.

Examples and documentation of xrun.Component should also make it clear on how to implement it.

Reply:
Here’s a working example https://github.com/gojekfarm/xtools/pull/7/files#diff-1df7c808f15fcba260509b1324e35b0a76caa546dda5ccb60bfba059d363057b